software/contrib/sequential_switch.py

```diff
@@ -0,0 +1,319 @@
+    def save(self):
+        """Save the current settings to persistent storage
+        """
+        state = {
+            "mode": self.mode,
+            "num_outputs": self.num_outputs
+        }
+        self.save_state_json(state)
+        
+    @classmethod
+    def display_name(cls):
+        return "Seq. Switch"
+    
+    def on_trigger(self):
+        """Handler for the rising edge of the input clock
+
+        Also used for manually advancing the output on a button press
+        """
+        # to save on clock cycles, don't use modular arithmetic
+        # instead just to integer math and handle roll-over manually
+        next_out = self.current_output
+        if self.mode == MODE_SEQUENTIAL:
+            next_out = next_out + 1
+        elif self.mode == MODE_REVERSE:
+            next_out = next_out - 1
+        elif self.mode == MODE_PINGPONG:
+            next_out = next_out + self.direction
+        else:
+            next_out = random.randint(0, self.num_outputs-1)
+            
+        if next_out < 0:
+            if self.mode == MODE_REVERSE:
+                next_out = self.num_outputs-1
+            else:
+                next_out = -next_out
+                self.direction = -self.direction
+        elif next_out >= self.num_outputs:
+            if self.mode == MODE_SEQUENTIAL:
+                next_out = 0
+            else:
+                next_out = self.num_outputs-2
+                self.direction = -self.direction
+                
+        self.current_output = next_out
+    
+    def main(self):
+        """The main loop
+
+        Connects event handlers for clock-in and button presses
+        and runs the main loop
+        """
+        
+        @din.handler
+        def on_rising_clock():
+            self.on_trigger()
+            
+        @b1.handler
+        def on_b1_press():
+            self.last_interaction_time = time.ticks_ms()
+            self.active_screen.on_button1()
+            
+        @b2.handler
+        def on_b2_press():
+            self.last_interaction_time = time.ticks_ms()
+        
+            if self.active_screen == self.switch_screen:
+                self.active_screen = self.menu_screen
+            else:
+                self.active_screen = self.switch_screen
+            
+        while True:
+            # keep the menu items sync'd with the left knob
+            self.menu_item = round(knob_rescale(k1, 0, len(self.menu_screen.menu_items)-1))
+            
+            # check if we've been idle for too long; if so, blank the screen
+            # to prevent burn-in
+            now = time.ticks_ms()
+            if time.ticks_diff(now, self.last_interaction_time) > SCREENSAVER_TIMEOUT_MS:
+                self.active_screen = self.screensaver
+            
+            # read the input and send it to the current output
+            # all other outputs should be zero
+            input_volts = ain.read_voltage()
+            for i in range(len(self.outputs)):
```

Comment:
[discussion] It occurs to me that you could eliminate this `for` loop if you set the voltages to 0 when switching outputs, instead of every time through the loop. That would trim this `while` loop down a bit and maybe reduce the bit crushing slightly. I'm not sure if the 'improvement' would actually be worth the complexity of setting the outputs in multiple places.

Reply:
It would speed up the while loop a little. But because on_trigger can be called asynchronously I'd need to add a mutex or a semaphore to control the current_output variable. Otherwise there's a chance that two outputs could get set to high if the timing of the ISR is _just_ right.

(I'm assuming the din events & button presses are triggering an interrupt at the chip level? I haven't looked that deeply into the code, but if I was writing this as a realtime systems assignment in university, that's how I probably would have implemented it.)

It's definitely something worth considering as a future optimization, but I'm not sure it's something I want to tackle and debug now.